### ISSUE TITLE
How to use gather_for_metrics for object detection models?

### ISSUE THREAD

**Issue by yann-rdgz** [huggingface/accelerate#2643]
### Reproduction

I used the `gather_for_metrics` function as follows:
```python
predictions, ground_truths = accelerator.gather_for_metrics((predictions, ground_truths))
```

And i've got the error:
```
accelerate.utils.operations.DistributedOperationException: Impossible to apply the desired operation due to inadequate shapes. All shapes on the devices must be valid.
```

* ground_truths are dictionaries of torch.tensor with keys: `boxes`, `labels`, `image_id`, `area`, `iscrowd` following pytorch conventions: https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html. 

* predictions are dictionaries of torch.tensor with `boxes`, `labels` and `scores` keys.

I use 3 gpus, and in each I have 120 dictionaries of predictions and ground truths, but as expected inside each dictionary the tensor size should vary from 0 to n bbox predictions/ground truths.
But during gather_predictions, the `verify_operation` decorator raises an  because all the tensor shapes inside the different dictionaries vary.

### Expected behavior

Have the possibility to gather complex objects like dictionaries of torch.tensor with different shapes!

Thank you for your help and for this amazing framework 🙏

**Comment by yann-rdgz** [comment]
currently to solve quick fix this issue I use this code:

```python
def gather_objects(objs, n_process):
    """Gather objects across all processes.

    This function gathers objects across all processes in a distributed setting.

    Parameters
    ----------
    objs : object
        The object to be gathered.

    Returns
    -------
    object
        The gathered object.
    """
    output_objects = [None for _ in range(n_process)]
    torch.distributed.all_gather_object(output_objects, objs)
    output_objects = [elem for obj in output_objects for elem in obj]

    return output_objects
```

with:
```python
ground_truth = gather_objects(ground_truth, n_process=accelerator.num_processes)
preds = gather_objects(preds, n_process=accelerator.num_processes)
```

**Comment by yann-rdgz** [comment]
solved in #2710
